ballista/rust/scheduler/src/state/task_manager.rs

```diff
@@ -608,29 +602,83 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> TaskManager<T, U>
-        self.metrics_collector
-            .set_pending_tasks_queue_size(result as f64);
+    pub fn increase_pending_queue_size(&self, num: usize) -> Result<()> {
+        if num != 0 {
+            let old_value = self.get_pending_task_queue_size();
+
+            if usize::MAX - old_value >= num {
+                let new_value = old_value + num;
+                match self.pending_task_queue_size.compare_exchange(
+                    old_value,
+                    new_value,
+                    AOrdering::AcqRel,
+                    AOrdering::Relaxed,
+                ) {
+                    Ok(_) => {
+                        self.metrics_collector
+                            .set_pending_tasks_queue_size(new_value as f64);
+                        debug!(
+                            "Pending queue size {} increased by {} to {}",
+                            old_value, num, new_value
+                        );
+                    }
+                    Err(_) => {
+                        return Err(BallistaError::Internal(format!(
+                            "Unable to increase pending queue size {} by {} to {}",
+                            old_value, num, new_value
+                        )))
+                    }
+                }
+            } else {
+                return Err(BallistaError::Internal(format!(
+                    "Refused to increase pending queue size {} by {}, wrap on overflow will happen", 
+                    old_value, num
+                )));
+            }
+        }
 
-        debug!("Pending queue size increased by {} to {}", num, result);
+        Ok(())
     }
 
-    pub fn decrease_pending_queue_size(&self, num: usize) {
-        if num > 0 {
-            let result = self
-                .pending_task_queue_size
-                .fetch_sub(num, AOrdering::Relaxed);
-            self.metrics_collector
-                .set_pending_tasks_queue_size(result as f64);
-
-            debug!("Pending queue size decreased by {} to {}", num, result);
+    pub fn decrease_pending_queue_size(&self, num: usize) -> Result<()> {
+        if num != 0 {
+            let old_value = self.get_pending_task_queue_size();
+
+            if old_value >= num {
+                let new_value = old_value - num;
+                match self.pending_task_queue_size.compare_exchange(
+                    old_value,
+                    new_value,
+                    AOrdering::AcqRel,
+                    AOrdering::Relaxed,
+                ) {
+                    Ok(_) => {
+                        self.metrics_collector
+                            .set_pending_tasks_queue_size(new_value as f64);
+                        debug!(
+                            "Pending queue size {} decreased by {} to {}",
+                            old_value, num, new_value
+                        );
+                    }
+                    Err(_) => {
+                        return Err(BallistaError::Internal(format!(
```

Comment:
I suppose we should fail here as we ended up in an inconsistent state, but it's still arguable